src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala

```diff
@@ -18,97 +19,149 @@ import scala.util.control.NonFatal
       }
     }
   }
 }
+
 /**
   * Processes durable operations on tasks by
   *
-  * * storing the updated tasks in the task repository
-  * * informing the taskTracker actor of the latest task state
+  * * storing the updated tasks in or removing them from the task repository
+  * * informing interested parts of Marathon by performing the configured [[TaskUpdateStep]]s.
   */
 private[tracker] class TaskOpProcessorImpl(
     taskTrackerRef: ActorRef,
     repo: TaskRepository,
-    statusUpdateActionResolver: StatusUpdateActionResolver) extends TaskOpProcessor {
+    stateOpResolver: TaskStateOpResolver,
+    steps: Seq[TaskUpdateStep],
+    metrics: Metrics) extends TaskOpProcessor {
+  import TaskOpProcessor._
+
   private[this] val log = LoggerFactory.getLogger(getClass)
 
-  import TaskOpProcessor._
+  private[this] val stepTimers: Map[String, Timer] = steps.map { step =>
+    step.name -> metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, s"step-${step.name}"))
+  }.toMap
 
-  override def process(op: Operation)(implicit ec: ExecutionContext): Future[Unit] = {
-    op.action match {
+  log.info("Started TaskOpProcessor with steps:\n{}", steps.map(step => s"* ${step.name}").mkString("\n"))
 
-      case Action.Update(task) =>
-        // Used for a create or as a result from a UpdateStatus action.
-        // The update is propagated to the taskTracker which in turn informs the sender about the success (see Ack).
-        val marathonTask = TaskSerializer.toProto(task)
-        repo.store(marathonTask).map { _ =>
-          taskTrackerRef ! TaskTrackerActor.TaskUpdated(task, TaskTrackerActor.Ack(op.sender))
-        }.recoverWith(tryToRecover(op)(expectedTaskState = Some(task)))
+  override def process(op: Operation)(implicit ec: ExecutionContext): Future[Unit] = {
+    val stateChange = stateOpResolver.resolve(op.stateOp)
 
-      case Action.Expunge =>
+    stateChange.flatMap {
+      case change: TaskStateChange.Expunge =>
         // Used for task termination or as a result from a UpdateStatus action.
         // The expunge is propagated to the taskTracker which in turn informs the sender about the success (see Ack).
         repo.expunge(op.taskId.idString).map { _ =>
-          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.taskId, TaskTrackerActor.Ack(op.sender))
-        }.recoverWith(tryToRecover(op)(expectedTaskState = None))
-
-      case Action.UpdateStatus(status) =>
-        statusUpdateActionResolver.resolve(op.taskId, status).flatMap { action: Action =>
-          // Since this action is mapped to another action, we delegate the responsibility to inform
-          // the sender to that other action.
-          process(op.copy(action = action))
-        }
+          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.taskId, TaskTrackerActor.Ack(op.sender, change))
+        }.flatMap { _ =>
+          notifyOthers(op.stateOp, change)
```

Comment:
As discussed, if someone outside of this package wants to update a task, the related future should only finish after the task tracker was updated and all steps executed sequentially for this update.

The steps themselves should only be executed after the TaskTracker has processed the related change. Therefore, it probably makes sense to call the step logic from the TaskTracker after it gets a change.